WIP,ENH: Allow `copy=np.never_copy` for array creation and reshape

seberg opened the pull request:
This morphed into the more general addition of `np.never_copy` (`"never"` or similar are problematic, because currently `np.array(..., copy="never")` would use `bool("never")` doing the exact opposite.

Things to do:
 - [x] Check if buffer/memoryview paths are allowed to make a copy, if they are, do not allow them here. (They do not.)
 - Add tests for:
     - [x] `np.array`
     - [ ] `arr.astype`
     - [ ] Deprecation/argument parsing errors
     - [ ] `np.reshape` function (and a few for attribute)
---

The current version includes a deprection for some bool conversions, to only allow integers at least, that can be split off probably.

seberg committed:
ENH: Allow to force nocopy/copy reshape in python

Adds a `copy={True, False, None}` kwarg to reshape on the python
side. Uses a hardcoded 1, 0, -1 flag right now. Could be exposed
on the C-side with a better flag.
Could also be used for `arr.shape = (...)` to raise the error
before doing the copy.

closes gh-11884

eric-wieser commented on doc/release/1.16.0-notes.rst:
```diff
@@ -106,6 +106,13 @@ details.
+--------------------------------------------------------------
+The reshape function in numpy will normally return a view but create a new
+array if necessary. A new ``copy`` keyword argument can toggle this behaviour.
+``copy=True`` will always copy the data, while ``copy=False`` will raise an
```
This is not consistent with the meaning of `copy=False` elsewhere. Right now, we have:

* `copy = True` - always copy
* `copy = False` - copy only if necessary

I think we should indicate never copy with `"never"`, `None`, or `"raise"`

shoyer replied:
Yes, I suppose we should probably handle the `copy` argument consistently with `np.array()`, for which `copy=False` means "only copy if necessary".

I like `copy='never'`

seberg replied:
So, I could just implement that on the C-Side (and also for reshape). But, for all exposed python functions that currently have the flag (mostly `np.asarray`, but `.astype` probably as well), the only way that does not require at least 2 years of waiting would be to create a `np.nevercopy` singleton, which would have to be supported indefinitely…

seberg replied:
Or, we do a funny middle way: Actually hard deprecate/remove the specific "never" string (who would use it?! :)) now. Just to speed things up. Should still have at least one version where "never" will simply break though.

seberg replied:
So `bool(None) is False` which is really strange currently, because it does not give the default of course. Anyway, doesn't matter. The only possible thing would be to ignore it for all but `reshape` or create a singleton and deprecate as much as possible

seberg commented:
Had the enum first, will put it in again (not soon maybe). Good point
about `copy=False` already meaning copy-if-necessary in some sense. Had
thought about strings but that is a very compelling reason!

seberg commented:
"never" sounds good, I think False might be good to avoid just to make
things more clear. One thing I just realized is that the input might be
converted to an array before with "`copy=False`". Possibly that is just
OK, but should think about it a bit.

shoyer commented:
Well, if we add copy=‘never’ to np.array() you could use it there, too :).
On Thu, Sep 13, 2018 at 11:41 PM Sebastian Berg <notifications@github.com>
wrote:

> "never" sounds good, I think False might be good to avoid just to make
> things more clear. One thing I just realized is that the input might be
> converted to an array before with "`copy=False`". Possibly that is just
> OK, but should think about it a bit.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/11897#issuecomment-421247387>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1s_3l65spuTh15OPOSKl2BeQmNdaks5ua0-LgaJpZM4Wd3hU>
> .
>

shoyer commented:
> The main problem would being certain whether `__array__` gives a new array or not

Do we actually call `__array__` for numpy arrays?

I would be OK treating every non-NumPy as a copy for the purposes of `np.array(..., copy='never')`.

seberg commented:
I looked into that change a bit. The code change seems pretty OK (although have to tracking down similar copy parameters elsewhere, such as in `astype` would be nice.

There is a major problem though, on older numpy  "never" would be interpreted as a boolean True, so it would do the exact opposite of the intention.

mhvk commented:
Might have to do `copy=None`?

mhvk commented:
Which is a bad suggestion - `None` might be some default one would set; it is not descriptive enough. Sorry for the noise.

seberg committed:
WIP: Implement nocopy `np.array` and `arr.astype`, also deprecate odd bool casting.

This will have to be split into multiple commits with better messages.

seberg committed:
WIP: Fixup the shape code a bit and create _wrapit for reshape

Reshape has to be a bit careful about the copies now after all...

seberg committed:
WIP,BUG: Fix error returns in ndarray creation for no-copy

seberg committed:
WIP,BUG: Debug the fast (and maybe some slow) paths, should work now.

seberg committed:
WIP,TST: Fix the new tests for np.never_copy

More are needed of course...

seberg commented:
If we actually consider this, should we go through an NEP or just the list?

shoyer commented:
I think the mailing list is fine for this. As a general rule, if you can
describe the design in a few short paragraphs then you probably don't need
a NEP.

On Fri, Dec 21, 2018 at 9:46 AM Sebastian Berg <notifications@github.com>
wrote:

> If we actually consider this, should we go through an NEP or just the list?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/11897#issuecomment-449453377>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1s1dR1PHFwnDqzg9ih7pcbvAkvIBks5u7R5qgaJpZM4Wd3hU>
> .
>

mhvk reviewed (commented):
Had a bit of a review (not quite complete), with some small comments.

seberg commented on numpy/_globals.py:
```diff
@@ -79,3 +80,35 @@ def __repr__(self):
+            cls.__instance = super(_NeverCopyType, cls).__new__(cls)
+        return cls.__instance
+
+    # needed for python 2 to preserve identity through a pickle
```
Yeah, you are right. thanks for the comments. It is still in a bit of an alpha stage overall, send an email to the list for now (but it seems to hang).

seberg commented on numpy/_globals.py:
```diff
@@ -79,3 +80,35 @@ def __repr__(self):
+    __instance = None
+    def __new__(cls):
+        # ensure that only one instance exists
+        if not cls.__instance:
```
Hmm, I copied this from above and it seems to work `bool(np.never_copy)` does raise the error for me, so I am not sure I understand. But maybe this can be much simpler anyway, since the pickle stuff is not necessary.

seberg replied:
Ohhhh, sorry, that was silly.

seberg commented on numpy/core/include/numpy/ndarraytypes.h:
```diff
@@ -868,6 +874,9 @@ typedef int (PyArray_FinalizeFunc)(PyArrayObject *, PyObject *);
  */
 #define NPY_ARRAY_WRITEABLE       0x0400
 
+/*      NPY_ARRAY_ENSURENOCOPY    0x0800 */
```
OOps, it seems there is alrady a flag with 0x0800, even if it is almost unused (and probably never collides).

mhvk commented:
Larger comments:

1. I'm still wondering about the implementation with the `np.never_copy` singleton. I understand the worry of using `copy='never'` as on older numpy it would evaluates to `copy=True`, which is the opposite of what is wanted. But now if anybody uses it the code just breaks on older numpy as `np.never_copy` doesn't exist. It is not obvious to me that is better... Since both would require people to make parts of their code version-dependent, perhaps just a warning in the docs is good enough? Is `None` better after all?

2. Should be go ahead and accept strings for `copy` in `reshape`? E.g., `{'never', 'atwill', 'always'}`

seberg commented on numpy/core/fromnumeric.py:
```diff
@@ -293,7 +312,11 @@ def reshape(a, newshape, order='C'):
-    return _wrapfunc(a, 'reshape', newshape, order=order)
+    # Since it is hard to tell if `np.array` had to copy, the copy is only
+    # forced during reshape (but no-copy forced also for `np.array`).
+    return _wrapfunc_copy(
```
I doubt introducing this `_wrapfunc_copy` is the nicest way to handle it, will think about it, but maybe someone has a nice idea.

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -764,6 +809,86 @@ PyArray_CastingConverter(PyObject *obj, NPY_CASTING *casting)
+        return NPY_SUCCEED;
+    }
+
+    /*
```
Did, just realized if it fails completely it may give the warning and an error. On the other hand, a failure here is very unlikely.

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1582,6 +1609,12 @@ PyArray_GetArrayParamsFromObject(PyObject *op,
             return -1;
         }
+        if (no_copy_allowed) {
+            PyErr_SetString(PyExc_ValueError,
```
Yes, definitely right. Quick copy paste :).

seberg replied:
Hmmm, I guess a static is good enough. Do we use global ones for this, since it is needed in two functions.

seberg commented:
Yeah, we could allow those strings (or similar ones) for `reshape` easily (another conversion func in C, but that is OK), which would be nice.

I am not sure that `"never"` is not an option also for `np.array`. It is a change in behaviour, but I am indeed more worried about code silently doing the wrong thing when written for a newer numpy and run on an older. For packages that is likely OK, but it is a possible silent breakage. I would always aim to allow it eventually, but it seems to me it probably needs one release deprecation and then at least one release of breakage.

seberg committed:
Small fixes/review comment fixes, also fix small refcount issue

And move the flag to the end, was thrown off by a left out flag
which is used elsewhere only.

Not quite all of the current review comments incorporated.

shoyer commented:
> I'm still wondering about the implementation with the `np.never_copy` singleton. I understand the worry of using `copy='never'` as on older numpy it would evaluates to `copy=True`, which is the opposite of what is wanted. But now if anybody uses it the code just breaks on older numpy as `np.never_copy` doesn't exist. It is not obvious to me that is better... Since both would require people to make parts of their code version-dependent, perhaps just a warning in the docs is good enough? Is `None` better after all?

I think it's definitely best to use `np.never_copy` and break loudly when someone tries to use this. The alternative of silent failure (i.e., silently copying in some cases) is not really an acceptable option.

If someone needs their code to run with older numpy, that's fine, they can hold off on using `np.never_copy`.

mhvk commented:
OK, that makes sense. In fact `copy=getattr(np, 'never_copy', False)` is pretty short and clear.

A variant on the implementation might be to have a `CopyFlag` class which it initialized with its string (with `__bool__` returning `False`, `True`, or raising an exception as appropriate (it would then have 3 instances). Though not sure this really helps with avoiding `np.never_copy`. Probably just overthinking.

seberg committed:
Remve assert message

We warn about it, it is not too useful and had it the wrong way
around anway

seberg committed:
Make argument NoValue (hack inside func right now)

Also tiny fixup for assert in c code and fix for array function
dispatcher.

seberg committed:
BUG: Fix dispatch, segfault (incorrect logic) and incorrect value passed bugs

seberg committed:
TST: Add new tests for the copyflag behaviour

Some tests still missing, mostly with respect to the memory order.

seberg committed:
DOC: Tiny changes to the docs/comments

seberg committed:
TST: Add tests for the mismatching order

seberg committed:
BUG: Fix wrong output type

seberg committed:
STY: start new line after long declaration

seberg committed:
TST: add basic test for copyflag to ndarray.astype

seberg committed:
TST: Add more tests, mostly for astype

godaygo commented:
Hi all! Reading the discussion on this issue, I remembered my first experience with the `copy` argument - I thought that using `copy=True` would always be a copy, and `copy=False` meant what is now proposed as `np.never_copy`. But reality turned out to be different. It was not so obvious why in some cases it was necessary to make a copy and in some not, before understanding how array is represented internally. But in any case, the semantics of `False` seemed strange to me (but there was no other option).

Now when changes are proposed, I think that at least it is good to discuss the posibility to make the switch in semantics as follows:
`copy=True` - always copy
`copy=False` - never copy, like the proposed `np.never_copy`.
`copy= np.if_needed` - like the current `False`.

As for me this way will be more newcomer friendly and semantically more correct. I'm not very familiar with the usual way of making such changes in `numpy`, so I apologize if my vision is completely wrong:)

seberg commented:
@godaygo it is not unreasonable, but there is a *lot* of usage for `np.array` and `ndarray.astype`. And in almost all cases it does exactly what it is supposed to. So, we would have to just do it in documentation for a long time. And it may be years before we can even do a warning...

godaygo commented:
@seberg you know better than me :) 

I was thinking about the same procedure that was done with `==` or fancy indexing, with pointing to the new way of doing right things in exception or warning. Also the `copy=getattr(np, 'if_needed', False)` better suits here. 

I understand that I am proposing some major change and it is usually a headache for depending projects. However I think this topic is worth discussing in some more visible place like mailing-list. 

Each long-living project acquires its own peculiarities - where `False` is not `False`, and `np.never_copy` is actually `False`, but cumulatively these mental discrepancies complicate the learning process and at some begin to annoy.

mhvk commented:
@godaygo - I think starting from scratch one would definitely want to use `copy=False` for never copy. The problem is indeed back compatibility, which means that, at best, if someone uses `copy=False` and a copy is needed, we could warn that behaviour will change. But that also means that even in the next version of numpy it would not be possible to insist that no copy is made using `copy=False`, i.e., we would still need a `np.never_copy` (or some other special version of `False`). In that sense, for this PR I don't think anything changes.

Still, it may be worth considering whether we should try to slowly move in the direction of having `copy=False` mean never copy.

godaygo commented:
@mhvk I do not know what is the status and attitude to the use of enums in numpy (I don't see any usage of them in codebase) but for this particular case it seems to me like a best fit, with the forward soft deprecation of current `False` behavior and a switch maybe in 2.x.

seberg commented:
@eric-wieser, thanks for looking through, great points. Personally, I would argue those all cap names are indeed only a semi conscious choice since they just reflect the `.h` files and were written by people familiar with C and those headers. And quite frankly, a huge part of our audience are people who take a 2 week course and then start fiddling with code. I see right now three alternatives:

 1. Be a bit more convincing that this is not just a niche feature, offsetting the annoyance of adding a new global name. While I work with huge video data, I cannot say I need it too much, although it would be nice just to be sure/clear!
 2. We use `ensure_view` or  `force_view` as an alternative new keyword argument (more kwargs, but in the code there will only ever be one of the two). This is a bit of a burden on duck arrays for the `.astype` method, which already has `copy` (`reshape` has no argument and `array/asarray` is not dispatched).
 3. We discuss if we should allow a more general `flags` methodology. The work is not an issue, the biggest problem I see is that it will create two ways to spell the same thing, which can be confusing.  (if someone knows python libs doing this off the top of the head, could be interesting)

eric-wieser commented:
> the biggest problem I see is that it will create two ways to spell the same thing

Could you elaborate on the two ways you were describing here?

seberg commented:
I guess I thought of the flags on the C-level, or in `nditer` (which is basically C-level). I.e. be forward looking if we must and actually do use `flags="never_copy"` (or whatever), so that you have a kwarg you could expand if you really wanted to.

In any case, I doubt its a good idea. I think we have to add the name, I would love to transition to or use a string, but not sure that can work (and we need to at least give a loud warning, as well as assume that *nobody* would ever have passed `"never_copy"` to mean `bool("never_copy")` before, which is probably a good assumption, but also a pretty weird one.)

IIRC this stalled because people aired opinions that it is too niche to add a new top-level name. I don't mind if we make it not toplevel (but the missing prior art is again annoying). And maybe the new discussion can give more reasons to just move forward.

eric-wieser commented:
Mailing list posts for those finding themselves here:

* [_"Add guaranteed no-copy to array creation and reshape?"_](http://numpy-discussion.10968.n7.nabble.com/Add-guaranteed-no-copy-to-array-creation-and-reshape-td46567.html#a46604)
* [_"Proposal: add `force=` or `copy=` kwarg to `__array__` interface"_](http://numpy-discussion.10968.n7.nabble.com/Proposal-add-force-or-copy-kwarg-to-array-interface-td48051.html)

eric-wieser commented:
Some other spellings that could be considered (alongside the ones already suggested above):

1. `copy='never'` (risky)
1. `copy=np.never_copy` (too namespace-polluting)
1. `copy=np.NEVER_COPY` (like 2 but marginally less so)
1. `copy=np.copy.never` (weird and not a general solution, but not namespace polluting)
1. some new namespace:
  1. `copy=np.arg.never_copy`  (`np.arg` might be expected to mean "complex argument"
  1. `copy=np.args.never_copy`
  1. `copy=np.opt.never_copy`
  1. `copy=np.opts.never_copy`
  1. `copy=np.flag.never_copy`
  1. `copy=np.flags.never_copy`
  If we took this approach, we could put `leave_wrapped` (#13105) and possible even `np._NoValue` in the same place

charris commented:
@seberg Needs rebase.

czgdp1807 commented:
I would like to work on continuing the work here. I tried to merge `main` branch into `nocopy_reshape`. It seems like there are a lot of conflicts which cannot be resolved easily. There are many updates since this PR which should be taken into account to support the never copy feature. Also, the discussions here on the value of `copy` argument seem to be unresolved.

I think this is a trade off between the meaning of copy and compatibility. Now, if we keep a global attribute within the numpy module (such as `np.never_copy`), then either it cannot take any non-boolean value due to the API of array defined [here](https://github.com/numpy/numpy/blob/6790873334b143117f4e8d1f515def8c7fdeb9fb/numpy/__init__.pyi#L3333-L3341) or we should change the array API (possibly an extension to accept non-boolean values in `copy` argument). The former approach will break the current code since, an error will be raised whenever copy cannot be avoided. However, the meaning of `False` will hold that is if the user has passed it as `False` then copy will never be made as is the intention. It will be easy to do since we will be required to just raise the error whenever `copy` is `False` but a copy is being made. In the second approach we can make a new attribute in the global namespace as is done in this PR and then raise an error whenever `copy` holds that value and a copy is being made. 

The question is what should be the value of `copy` to ensure that no copy is being made? One way is string, but in the above comment it is said to be risky. Well, as far as I can understand, the reason behind this might be, that it is fragile, one can pass `'Never'`, `'never'` (different cases). I would like to know if there is any other reason to not use strings as it is an easy way without introducing new attributes in the namespace.  Adding attributes is good but still leads to pollution and unwanted stuff. Using something like, `np.flags.never_copy` is better as `never_copy` is a flag used as a signal.

I would lik

seberg commented:
Yeah, the merge conflicts might make it easier to copy the good parts rather than trying to resolve the conflict itself.  Large chunks will probably still be fine, but quite a bit of things were modified here changed in the past year.  I am happy to give pointers or discuss this though.

To move this forward, the main thing necessary is to make a decision of what `"never copy"` should look like exactly and how it should be passed.  Once the API is settled, I expect we can push things forward reasonably well (although I don't recall how tricky the C-API part was here).

> The former approach will break the current code since, an error will be raised whenever copy cannot be avoided.

Yes, breaking code that tries to use no-copy on an old NumPy version should just raise an exception.  If the user wants to use it only when available, they will need to use the pattern:

```python
never_copy = np.never_copy if hasattr(np, "never_copy") else False
np.array(..., copy=never_copy)
```

>  Well, as far as I can understand, the reason behind this might be, that it is fragile, one can pass 'Never', 'never' (different cases)

The reason that we consider it risky is backward compatibility: `bool("never")` is `True`.  That means if you run code such as `np.array(..., copy="never")` on current NumPy, you will actually *force* a copy, rather than the opposite (ensure no copy)!  So it is risky in the sense of allowing for big confusions when code that looks perfectly fine runs on older NumPy versions.

czgdp1807 commented:
I thought of something about extension of API so that all of the options, `True`, `False` and `never_copy` are accommodated. Let's say we change the type of `copy` argument from `bool` to `int`. Now, if user passes, `True` then it will be converted to `1`, `False` to `0` and say we keep `never_copy` as `2`. Now, we can check if the `copy` has any of the values and decide if an error should be raised or not in case a copy is being made. For easy coding, we can have some names for `0`, `1` and `2` in the global namespace. Say, `np.COPY_IF_NEEDED = 0`, `np.ALWAYS_COPY = 1`, `np.NEVER_COPY = 2`. Anywhere inside the numpy codebase, we can add checks like, `copy == np.NEVER_COPY` or `copy == np.COPY_IF_NEEDED`, etc. One advantage of this approach is that existing user code will not be required to change since the API is extended and not deprecated. We can just cast the user input for copy to `int` by, `int(copy)`. One disadvantage is that anything like, `copy = None` will raise an error. Do we accept `None` as a value for `copy`?

shoyer commented:
If we're going to go this route, I would just make an actual Python enum
with three value, e.g., np.Copy with attributes ALWAYS, NEVER and IF_NEEDED.

On Wed, Jun 2, 2021 at 7:44 PM Gagandeep Singh ***@***.***>
wrote:

> I thought of something about extension of API so that all of the options,
> True, False and never_copy are accommodated. Let's say we change the type
> of copy argument from bool to int. Now, if user passes, True then it will
> be converted to 1, False to 0 and say we keep never_copy as 2. Now, we
> can check if the copy has any of the values and decide if an error should
> be raised or not in case a copy is being made. For easy coding, we can have
> some names for 0, 1 and 2 in the global namespace. Say, np.COPY_IF_NEEDED
> = 0, np.ALWAYS_COPY = 1, np.NEVER_COPY = 2. Anywhere inside the numpy
> codebase, we can add checks like, copy == np.NEVER_COPY or copy ==
> np.COPY_IF_NEEDED, etc. One advantage of this approach is that existing
> user code will not be required to change since the API is extended and not
> deprecated. We can just cast the user input for copy to int by, int(copy).
> One disadvantage is that anything like, copy = None will raise an error.
> Do we accept None as a value for copy?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/11897#issuecomment-853518692>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJJFVSNLQPTD4MY47I6UJTTQ3UA5ANCNFSM4FTXPBKA>
> .
>

czgdp1807 commented:
> If we're going to go this route, I would just make an actual Python enum
with three value, e.g., np.Copy with attributes ALWAYS, NEVER and IF_NEEDED.

Yes, that would be much better. Though, are there any disadvantages of using `int` type `copy` argument?

shoyer commented:
> > If we're going to go this route, I would just make an actual Python enum
> > with three value, e.g., np.Copy with attributes ALWAYS, NEVER and IF_NEEDED.
> 
> Yes, that would be much better. Though, are there any disadvantages of using `int` type for `copy` argument?

The main advantage of an enum is that we can overload it so that casting it to a boolean raises an error, e.g.,
```
import enum

class Copy(enum.Enum):
  NEVER = 0
  ALWAYS = 1
  IF_NEEDED = 2

  def __bool__(self):
    raise TypeError(f"{self} is neither true nor false")

bool(Copy.NEVER)
# TypeError: Copy.NEVER is neither true nor false
```

With integers, there is some risk that someone could pass the value into a function that expects a boolean and it could be silently misinterpreted. This could come up with alternative implementations of NumPy's API via `__array_function__`, for example.

czgdp1807 commented:
I see. Its a good approach. So, `copy` will be of type `Copy` instead of `bool`. I think under `Copy` enum, `NEVER = 2`, and `IF_NEEDED = 0` so that code using current API doesn't break as `False` is same as 0 so it should map to `IF_NEEDED`.

seberg commented:
Making it an `Enum` seems right to me as well.  The other question is where to put such an Enum. In the main namespace or a new one (or both).

It would be amazing to just point to some prior art. Python uses Enum's a fair bit now and I think always puts them into the same namespace. But they dot have the problem of having the huge namespace that NumPy has.

If our only/main concern is `__array_function__` compatibility, we could in theory normalize the argument there.  That might make sense if we have a long term plan of also allowing using `copy="if_needed"`, as a string in the long run.

(Bike-shedding for later:  I think the enum needs an extra word in its name, something like: `…Options`, `…Mode`, or `…Flags`.  In this case, `CopyMode` seems right.)

shoyer commented:
> (Bike-shedding for later: I think the enum needs an extra word in its name, something like: `…Options`, `…Mode`, or `…Flags`. In this case, `CopyMode` seems right.)

`np.CopyMode.NEVER` looks pretty good to me. I don't think it's worth the trouble of a separate namespace just for this enum.

czgdp1807 commented:
> That might make sense if we have a long term plan of also allowing using copy="if_needed", as a string in the long run.

One problem with using string as argument is the user experience, in the sense that values of copy are not consistent (not of same type, `True`/`False`/"never_copy"/"if_needed"). One may have to remember what string is exactly expected and considering large code bases using numpy as a dependency, it might get difficult to make the change everywhere. However an enum will keep things consistent and probably the API won't break as well if we design the enum carefully. Just a bunch of thoughts. 

So, if all is good then I will make a new PR taking inspiration from the changes in these and we can carry on the discussions there as well while trying out different things. Let me know if it works. Thanks.

seberg commented:
Works for me, thanks.

seberg commented:
Closing, since most of this is conflicting (largely due to being addressed in gh-19173).  The approach to `reshape` is probably still valid though.
